File: pe-core/src/main/java/org/saar/core/mesh/lod/LodMesh.java
package org.saar.core.mesh.lod;

import org.saar.core.mesh.Mesh;

public class LodMesh implements Mesh {

    private LodMeshHelper helper = LodMeshHelper.empty();

    public void addMesh(Mesh mesh) {
        this.helper = this.helper.addMesh(mesh);
    }

    public void removeMesh(Mesh mesh) {
        this.helper = this.helper.removeMesh(mesh);
    }

    public LevelOfDetail getLod() {
        return this.helper.getLod();
    }

    @Override
    public void draw() {
        this.helper.draw();
    }

    @Override
    public void delete() {
        this.helper.delete();
    }
}

File: pe-core/src/main/java/org/saar/core/mesh/lod/LodMesh.kt
package org.saar.core.mesh.lod

import org.saar.core.mesh.Mesh

class LodMesh(private val meshes: List<Mesh>) : Mesh {

    val lod = ClampedInt(0, this.meshes.size)

    override fun draw() {
        val lod = this.lod.get()
        val mesh = this.meshes[lod]
        mesh.draw()
    }

    override fun delete() = this.meshes.forEach(Mesh::delete)
}
